### PR TITLE
Switch to Drake's `AddSimIiwaDriver` and enable missing control modes for iiwa hardware

### PR DESCRIPTION
**Pull request by nepfaff**
So far, we ignored the iiwa driver `control_mode`. This PR uses the control mode when setting up LCM connections with the iiwa hardware. In particular, `iiwa.position` is no longer exported when the iiwa is in `torque_only` mode.

Note that this does not implement a simulated version for non-default (`position_and_torque`) control modes but adds a warning instead.

**Question:** This seems to be duplicating much of what [BuildIiwaControl](https://drake.mit.edu/doxygen_cxx/namespacedrake_1_1manipulation_1_1kuka__iiwa.html#ad3720de284f34533b4dbbaa0710637b4) is doing. Is there a reason for this?

- - -
This change is [Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293)

### CODE REVIEW

**Comment by nepfaff** [comment]
+@RussTedrake for review please. This change is breaking if someone specified `torque_only` mode in their scenario file (very unlikely).

**Review by RussTedrake — COMMENTED** [review]
I definitely think we should converge on `BuildIiwaControl` going forward. I think it's farther along now then when I started this.  Is it possible to switch to it now, instead of reimplementing things here?

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293#-:-NrDeT2J4r9HWuhpu08_:bst81o4)* status: 0 of 1 files reviewed, all discussions resolved

**Comment by nepfaff** [comment]
I talked with Jeremy, and we don't want to use `BuildIiwaControl` in its current state for hardware control. So this change is needed

**Comment by nepfaff** [comment]
We should be able to use Drake's [AddSimIiwaDriver](https://github.com/RobotLocomotion/drake/blob/master/manipulation/kuka_iiwa/build_iiwa_control.cc#L147) as proposed here. However, I'm struggling with the binding, so blocked until TRI is back in the office to help.

**Comment by RussTedrake** [comment]
I can likely help with the binding…?

**Review by nepfaff — COMMENTED** [review]
Jeremy did respond already :tada:
https://drakedevelopers.slack.com/archives/C3YB3EV5W/p1708824121399429

I can take this one. The sooner I get more familiar with Drake development, the better 

> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293#-:-NrT2FAj0QH8NLNEHpCA:b-rh9q9t)* status: 0 of 1 files reviewed, all discussions resolved (waiting on @RussTedrake)

**Comment by nepfaff** [comment]
I expect this to be fine once the Drake PR has landed and the Drake nightly version has been added here.

File: manipulation/station.py
Comment (nepfaff) [inline manipulation/station.py:590]: We must keep `controller_plant` alive per contract with `AddSimIiwaDriver`. Otherwise, this will result in a segmentation fault. I don't think there is a neat `keep_alive` solution on the C++ side here (`InverseDynamicsController` secretly uses a `keep_alive` trick and thus does not rely on its contract). I will think of a neat way of doing this.

Comment (nepfaff) [reply manipulation/station.py:591]: the pointer system takes care of this :tada:

**Comment by nepfaff** [comment]
This should be ready now s.t. being able to peg the correct drake version. It might have broken some of the manipulation course notebooks due to the new default port naming/ port presence:
- "feedforward_torque" -> "torque"
- `position_and_torque` (default) requires you to connect both "torque" and "position" ports. If you only want to use the "position" port, then you need to explicitly specify "position_only" in the scenario file.

**Comment by nepfaff** [comment]
Everything is passing locally now. Should be ready for review.

**Review by RussTedrake — CHANGES_REQUESTED** [review]
> Reviewed 2 of 7 files at r5, 22 of 23 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293)* status: all files reviewed, 6 unresolved discussions (waiting on @nepfaff)

___
*[`book/clutter/suction.ipynb` line 129 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQufXjEB1XN4c_WBxT:-NsQufXjEB1XN4c_WBxU:b2j5den) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/book/clutter/suction.ipynb#L129)):*
> ```
>     "model_drivers:\n",
>     "    iiwa: !IiwaDriver\n",
>     "        control_mode: position_only\n",
> ```

can't we have control_model default to position_only (or position_and_torque) so that the default is still valid? 
___
*[`book/pick/exercises/08_robot_painter.ipynb` line 142 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQvgoOCTnnFadvyAUD:-NsQvgoOCTnnFadvyAUE:b-txenhp) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/book/pick/exercises/08_robot_painter.ipynb#L142)):*
> ```
>     "            iiwa_position = builder.AddSystem(ConstantVectorSource(np.zeros(7)))\n",
>     "            builder.Connect(\n",
>     "                iiwa_position.get_output_port(),\n",
> ```

why is this connection required now when it wasn't before?
___
*[`book/trajectories/exercises/door_opening.ipynb` line 72 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQvpwB8akwUZQXtNNQ:-NsQvpwB8akwUZQXtNNR:bkc7urr) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/book/trajectories/exercises/door_opening.ipynb#L72)):*
> ```
>     "    builder.Connect(\n",
>     "        iiwa_position.get_output_port(), station.GetInputPort(\"iiwa.position\")\n",
>     "    )\n",
> ```

why is this connection required now when it wasn't before? (repeated so that we can zap if we change the root cause)
___
*[`book/trajectories/exercises/rrt_planning.ipynb` line 78 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQvrSS7VNq8uOc0Zr9:-NsQvrSS7VNq8uOc0ZrA:b-txenhp) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/book/trajectories/exercises/rrt_planning.ipynb#L78)):*
> ```
>     "        self.station.GetInputPort(\"iiwa.position\").FixValue(\n",
>     "            self.context_station, np.zeros(7)\n",
>     "        )\n",
> ```

why is this connection required now when it wasn't before?
___
*[`manipulation/station.py` line 67 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQw4FG0JWCc4wVeYjc:-NsQw4FG0JWCc4wVeYjd:b2x22yn) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/manipulation/station.py#L67)):*
> ```Python
>     VisualizationConfig,
>     ZeroForceDriver,
>     position_enabled,
> ```

i'm surprised these are in the global namespace..?  that seems wrong (on the Drake side?). I would think it would be e.g. IiwaControlMode.position_enabled?
___
*[`manipulation/station.py` line 581 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQwPdvE_khZiCSBfT_:-NsQwPdvE_khZiCSBfTa:bxs3jv8) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/manipulation/station.py#L581)):*
> ```Python
> 
>         # Make the plant for the iiwa controller to use.
>         # TODO: The current hardcoded implementaiton should be replaced with the
> ```

nit: type



_Code quote:_
```Python
implementaiton
```

**Comment by RussTedrake** [comment]
fyi -- i've just added you to the manipulation-solutions repo.  all of the notebooks here that are in an `exercises` subfolder are actually installed from the solutions repo, so also need to change upstream.  The recommendation is that you check out the manipulation-solutions repo into e.g. a `solutions` subdirectory of this repo (so that bazel runs those tests, too.

**Review by nepfaff — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293)* status: all files reviewed, 6 unresolved discussions (waiting on @RussTedrake)

___
*[`book/clutter/suction.ipynb` line 129 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQufXjEB1XN4c_WBxT:-NsT_v6WBf6I8MwKPHJa:blu6qu9) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/book/clutter/suction.ipynb#L129)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

can't we have control_model default to position_only (or position_and_torque) so that the default is still valid?
</blockquote></details>

It is not possible to keep the previous behavior where the position is required, and torque is optional. Simply based on Drake's upstream implementation that I agree with from a hardware perspective. See my comment above:

```
This should be ready now s.t. being able to peg the correct drake version. It might have broken some of the manipulation course notebooks due to the new default port naming/ port presence:
-   "feedforward\_torque" -> "torque"
-   `position_and_torque` (default) requires you to connect both "torque" and "position" ports. If you only want to use the "position" port, then you need to explicitly specify "position\_only" in the scenario file. 
```
___
*[`book/pick/exercises/08_robot_painter.ipynb` line 142 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQvgoOCTnnFadvyAUD:-NsTaJoC7WYTDw-61xiW:byr4ffv) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/book/pick/exercises/08_robot_painter.ipynb#L142)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

why is this connection required now when it wasn't before?
</blockquote></details>

see above
___
*[`book/trajectories/exercises/door_opening.ipynb` line 72 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQvpwB8akwUZQXtNNQ:-NsTaLRX68LWuTqlN8h8:byr4ffv) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/book/trajectories/exercises/door_opening.ipynb#L72)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

why is this connection required now when it wasn't before? (repeated so that we can zap if we change the root cause)
</blockquote></details>

see above
___
*[`book/trajectories/exercises/rrt_planning.ipynb` line 78 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQvrSS7VNq8uOc0Zr9:-NsTaLydEnx-uRFr2GVU:byr4ffv) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/book/trajectories/exercises/rrt_planning.ipynb#L78)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

why is this connection required now when it wasn't before?
</blockquote></details>

see above
___
*[`manipulation/station.py` line 67 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQw4FG0JWCc4wVeYjc:-NsTaMqmE6fjaPsOncot:b-du8uci) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/manipulation/station.py#L67)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i'm surprised these are in the global namespace..?  that seems wrong (on the Drake side?). I would think it would be e.g. IiwaControlMode.position_enabled?
</blockquote></details>

It does seem weird. We could complain about it, but we would need to keep this for backward compatibility. These were already exposed before I added bindings.

**Review by nepfaff — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293)* status: 24 of 25 files reviewed, 6 unresolved discussions (waiting on @RussTedrake)

___
*[`manipulation/station.py` line 1093 at r8](https://reviewable.io/reviews/RussTedrake/manipulation/293#-Ns_l9aR-H_BGwzIG2eJ:-Ns_l9aR-H_BGwzIG2eK:b-fojk3) ([raw file](https://github.com/RussTedrake/manipulation/blob/4fbd65e44b0d9a31cc106a931d22f7bd648f8045/manipulation/station.py#L1093)):*
> ```Python
>         builder.ExportOutput(
>             iiwa_status_receiver.get_torque_measured_output_port(),
>             model_instance_name + ".torque_measured",
> ```

It looks like for sim compatibility, we need to flip the LCM torque output ports. See https://github.com/RobotLocomotion/drake/pull/21077

Should we flip the sim or the hardware ones? I think flipping hardware would be more usable as it would work with Drake convention (positive acceleration corresponds to positive torque).

**Review by nepfaff — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293)* status: 24 of 25 files reviewed, 6 unresolved discussions (waiting on @RussTedrake)

___
*[`manipulation/station.py` line 1093 at r8](https://reviewable.io/reviews/RussTedrake/manipulation/293#-Ns_l9aR-H_BGwzIG2eJ:-Ns_rErx37KDg8-c8y4c:b-cdj6fv) ([raw file](https://github.com/RussTedrake/manipulation/blob/4fbd65e44b0d9a31cc106a931d22f7bd648f8045/manipulation/station.py#L1093)):*
<details><summary><i>Previously, nepfaff (Nicholas Pfaff) wrote…</i></summary><blockquote>

It looks like for sim compatibility, we need to flip the LCM torque output ports. See https://github.com/RobotLocomotion/drake/pull/21077

Should we flip the sim or the hardware ones? I think flipping hardware would be more usable as it would work with Drake convention (positive acceleration corresponds to positive torque).
</blockquote></details>

I added the change, will test it on Monday, and push after that. I mainly added this comment to point out the convention change (that should have no casualties this time) and block the PR from being merged until it is completed.

**Review by RussTedrake — APPROVED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @nepfaff)

___
*[`manipulation/station.py` line 67 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQw4FG0JWCc4wVeYjc:-NsiJu2T-NaeIRWzlCyG:b-xme2pr) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/manipulation/station.py#L67)):*
<details><summary><i>Previously, nepfaff (Nicholas Pfaff) wrote…</i></summary><blockquote>

It does seem weird. We could complain about it, but we would need to keep this for backward compatibility. These were already exposed before I added bindings.
</blockquote></details>

OK for the PR.  I would like to clean things up upstream.  Would you mind opening a PR?

**Comment by RussTedrake** [comment]
did you manage to make the updates in solutions for me, too?

**Comment by nepfaff** [comment]
> did you manage to make the updates in solutions for me, too?

As per slack, I never received the invitation. Would you mind re-sending it?

**Review by nepfaff — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293)* status: 24 of 25 files reviewed, all discussions resolved (waiting on @RussTedrake)

___
*[`manipulation/station.py` line 67 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQw4FG0JWCc4wVeYjc:-Nsiusa20TGv3GIB9l-o:bbxzrrh) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/manipulation/station.py#L67)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

OK for the PR.  I would like to clean things up upstream.  Would you mind opening a PR?
</blockquote></details>

Sure. I'm not quite sure how to deal with backward compatibility. Do you know of a PR that I could refer to?

**Review by RussTedrake — COMMENTED** [review]
it looks like you've got (accepted?) permissions.  https://github.com/RobotLocomotion/manipulation-solutions

> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293#-:-Nt_KF316beZdCudaWEJ:b-p62uj1)* status: 24 of 25 files reviewed, all discussions resolved (waiting on @nepfaff)

___
*[`manipulation/station.py` line 67 at r7](https://reviewable.io/reviews/RussTedrake/manipulation/293#-NsQw4FG0JWCc4wVeYjc:-Nt_KSil3NsHIJLpNqZN:bv2kis5) ([raw file](https://github.com/RussTedrake/manipulation/blob/bf641d5d6f47573ccea1bac6d9d32ff7cf5e4d22/manipulation/station.py#L67)):*
<details><summary><i>Previously, nepfaff (Nicholas Pfaff) wrote…</i></summary><blockquote>

Sure. I'm not quite sure how to deal with backward compatibility. Do you know of a PR that I could refer to?
</blockquote></details>

i would think that we would introduce the enum again in the proper namespace and deprecate the enum in the current namespace.  Any methods that use the old enum would be deprecated and an overload with the new enum would be created instead.

**Review by RussTedrake — COMMENTED** [review]
Should we merge this now, or wait?  (I've recently fixed the CI failures on mac, but by upgrading to sonoma; the monterey CI will continue to fail).

> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293#-:-Nt__80OBmYmTzPmAHtC:b2wavr5)* status: 24 of 25 files reviewed, all discussions resolved (waiting on @nepfaff)

**Comment by nepfaff** [comment]
> Should we merge this now, or wait? (I've recently fixed the CI failures on mac, but by upgrading to sonoma; the monterey CI will continue to fail).
> 
> > _[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293#-:-Nt__80OBmYmTzPmAHtC:b2wavr5)_ status: 24 of 25 files reviewed, all discussions resolved (waiting on @nepfaff)

Looks like I can access the solution repo. I just didn't get a notification and wasn't sure where it was located.
I should probably fix the solution repo before merging this. What is the usual workflow here? I have this on my backlog, but I'm quite swamped at the moment. I should be able to do it in two weeks.

**Review by RussTedrake — APPROVED** [review]
PR to solutions, then run the install script in solutions to copy the files over (with the solutions removed) into the main repo.

> Reviewed 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293#-:-Ntkh0x8E2f5GZz4l9q9:b-6h7v4k)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @nepfaff)

**Review by nepfaff — COMMENTED** [review]
Done. This should now be ready to merge after a final review.

> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293#-:-NujHavW4G_Savuup_n-:b30ppig)* status: 12 of 27 files reviewed, all discussions resolved (waiting on @RussTedrake)

**Review by RussTedrake — APPROVED** [review]
> Reviewed 15 of 15 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/293)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @nepfaff)

**Comment by RussTedrake** [comment]
The mac failure is, I think, not caused by you.  The pip failure is because I'll need to roll new pip wheels after this merges.

**Pull request merged**
